experimenter/manifesttool/fetch.py

```diff
@@ -0,0 +1,156 @@
+
+    try:
+        # We could operate against "main" for all these calls, but the repository
+        # state might change between subsequent calls. That would mean the fetched
+        # feature schemas could differ or not be present if they were removed in a
+        # subsequent commit.
+        rev = result.ref = hgmo_api.get_tip_rev(app_config.repo.name)
+        print(f"fetch-latest: {app_name}: tip is {rev}")
+
+        manifest_path = manifest_dir / app_config.slug / "experimenter.yaml"
+        print(f"fetch-latest: {app_name}: downloading experimenter.yaml")
+        hgmo_api.fetch_file(
+            app_config.repo.name,
+            app_config.experimenter_yaml_path,
+            rev,
+            manifest_dir / app_config.slug / "experimenter.yaml",
+        )
+
+        with manifest_path.open() as f:
+            raw_manifest = yaml.safe_load(f)
+            manifest = FeatureManifest.parse_obj(raw_manifest)
+
+        schema_dir = manifest_dir / app_config.slug / "schemas"
+        schema_dir.mkdir(exist_ok=True)
+
+        # Some features may re-use the same schemas, so we only have to fetch them
+        # once.
+        fetched_schemas = set()
+
+        for feature_slug, feature in manifest.__root__.items():
+            feature = feature.__root__
+
+            if feature.json_schema is not None:
+                if feature.json_schema.path in fetched_schemas:
+                    print(
+                        f"fetch-latest: {app_name}: already fetched schema for "
+                        f"feature {feature_slug} ({feature.json_schema.path})"
+                    )
+                    continue
+
+                print(
+                    f"fetch-latest: {app_name}: fetching schema for feature {feature_slug} "
+                    f"({feature.json_schema.path})"
+                )
+
+                schema_path = schema_dir.joinpath(*feature.json_schema.path.split("/"))
+                schema_path.parent.mkdir(exist_ok=True, parents=True)
+
+                hgmo_api.fetch_file(
+                    app_config.repo.name,
+                    feature.json_schema.path,
+                    rev,
+                    schema_path,
+                )
+
+                fetched_schemas.add(feature.json_schema.path)
+    except Exception as e:
+        print_exception(e, file=sys.stderr)
+        result.exc = e
+
+    return result
+
+
+def summarize_results(results: list[FetchResult]):
+    print("\n\nSUMMARY:\n")
+
+    if any(result.exc is None for result in results):
+        print("SUCCESS:\n")
+        for result in results:
+            if result.exc is None:
+                print(f"{result.app_name} at {result.ref_name} ({result.ref})")
+
+    print("")
+    if any(result.exc is not None for result in results):
+        print("FAILURES:\n")
+        for result in results:
+            if result.exc is not None:
+                print(f"{result.app_name} at {result.ref_name} ({result.ref})")
```

Comment:
Couldn't you just write this as

```py
        for result in results:
            if result.exc is None:
                print(f"SUCCESS: {result.app_name} at {result.ref_name} ({result.ref})")
            else:
                print(f"FAILED: {result.app_name} at {result.ref_name} ({result.ref})
```

Reply:
This groups all the successes and all the failures